cpp/arcticdb/util/stats_query.hpp

```diff
@@ -0,0 +1,118 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#pragma once
+#include <mutex>
+#include <memory>
+#include <vector>
+#include <list>
+#include <map>
+#include <atomic>
+#include <type_traits>
+#include <string>
+#include <ctime>
+#include <chrono>
+#include <fmt/format.h>
+
+namespace arcticdb::util::stats_query {
+using GroupableStats = std::vector<std::shared_ptr<std::pair<std::string, std::string>>>;
+
+
+// thread-global instance
+class StatsInstance { // this will be passed to folly worker threads
+public:
+    static std::shared_ptr<StatsInstance> instance();
+    static void copy_instance(std::shared_ptr<StatsInstance>& ptr);
+    static void pass_instance(std::shared_ptr<StatsInstance>&& ptr);
+
+    GroupableStats info_;
+private:
+    thread_local inline static std::shared_ptr<StatsInstance> instance_;
+};
+
+// process-global stats entry list
+class GroupableStat;
+using StatsOutputFormat = std::vector<std::map<std::string, std::string>>;
+class StatsQuery {
+public:
+    void reset_stats();
+    
+    template<typename T>
+    void add_stat(std::shared_ptr<StatsInstance>& stats_instance, const std::string& col_name, T&& value) {
+        std::lock_guard<std::mutex> lock(stats_mutex_);
+        stats.emplace_back(stats_instance->info_, std::make_pair(col_name, fmt::format("{}", std::forward<T>(value))));
+    }
+
+    StatsOutputFormat get_stats();
+    bool is_enabled();
+    void register_new_query_stat_tool();
+    void deregister_query_stat_tool();
+    static StatsQuery& instance();
+
+    // TODO: Support enable in the work of a particualr python thread only, need to 
+    // 1. Convert this to python_thread_local variable; Check Py_tss_t in python.h
+    // 2. Remove atomic
+    std::atomic<bool> stats_query_enabled = false;
+private:
+    std::atomic<int32_t> query_stat_tool_count = 0;
+    std::mutex stats_mutex_;
+    //TODO: Change to std::list<std::pair<GroupableStats, std::pair<std::string, std::variant<std::string, xxx>>> 
```

Comment:
This desperately needs some type aliases or structs pulling out. And should implement the TODO now if you want it.